### PR TITLE
Add flash_mla_prefill (Multi-head Latent Attention prefill) op support

### PR DESCRIPTION
**Pull request by hshahTT**
### Ticket
https://github.com/tenstorrent/tt-xla/issues/3884

### Problem description
tt-mlir has no support for Multi-head Latent Attention (MLA) prefill. This change adds an end-to-end lowering path for `ttnn.transformer.flash_mla_prefill`.

### What's changed
- **New `flash_mla_prefill` op** added to the TTNN dialect, mirroring `ttnn.transformer.flash_mla_prefill` semantics (separate Q/K head dim vs. V/output `head_dim_v`, optional `value`/`attention_mask`, `is_causal`, and `scale`), with verifiers enforcing the shape and causal/mask constraints.
- **StableHLO → TTIR conversion: **  converts the `stablehlo.custom_call` op generated by vLLM into a `ttcore.composite` op.
- **TTIR → TTNN conversion** lowering the op to the TTNN dialect.
- **EmitC and EmitPy emitters** for generating C++/Python code for the op.
- **Flatbuffer schema + serialization**: added the op to `transformer.fbs`/`program.fbs` and `TTNNToFlatbuffer`.
- **Runtime implementation** (`flash_mla_prefill.cpp/.h`) wired into the program executor and runtime.
- **OpModel support**: op model interface and `TTNNOpModel` entries for constraint/perf queries.
- **TTNN workarounds**: a dtype-aware bf16 cast for the op's operands — Q/K/V and the output are cast to bf16 only when not already in a tt-metal SDPA-supported dtype (`bf16`/`bfp8_b`/`bfp4_b`)
- **Custom sharding rules** (`RegisterCustomShardingRule.cpp`) for MLA prefill.
- **Builder + golden support**: added SHLO builder support and golden mapping for the op.
- **Tests** across the stack: TTIR positive/negative, StableHLO conversion, TTNN workarounds, EmitC/EmitPy, runtime silicon, and Python builder tests.

### Checklist
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

**Review by jameszianxuTT — APPROVED** [review]
Runtime op impl LGTM

File: lib/Dialect/TTNN/Interfaces/TTNNOpModelInterface.cpp [hunk 1]
```diff
@@ -2403,6 +2403,78 @@ llvm::Expected<size_t> ScaledDotProductAttentionOp::getOpRuntime(
...
+      getValue() ? std::make_optional(getValue().getType().getShape())
+                 : std::nullopt;
+  const std::optional<TTNNLayoutAttr> valueLayout =
+      getValue() ? std::make_optional(inputs[idx++]) : std::nullopt;
+  const std::optional<llvm::ArrayRef<int64_t>> attentionMaskShape =
+      getAttentionMask()
+          ? std::make_optional(getAttentionMask().getType().getShape())
+          : std::nullopt;
+  const std::optional<TTNNLayoutAttr> attentionMaskLayout =
+      getAttentionMask() ? std::make_optional(inputs[idx++]) : std::nullopt;
```
Comment (sgholamiTT) [inline lib/Dialect/TTNN/Interfaces/TTNNOpModelInterface.cpp]: Can you please refactor this code into a free func `unpackFlashMlaPrefillArgs` that can be called from both `getOpRuntime` and `getOpConstraint` ?

**Review by sgholamiTT — APPROVED** [review]
OpModel changes LGTM 👍🏻 
Please add tests in `TestOpModelInterface` and `TestOpModelLib` for this Op. Tnx.

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp [hunk 2]
```diff
@@ -8614,6 +8614,131 @@ class StableHLOToTTIRScaledDotProductAttentionOpConversionPattern
...
+    // Required: head_dim_v (string -> uint32).
+    auto headDimVStringAttr =
+        frontendAttributes.getAs<mlir::StringAttr>("head_dim_v");
+    if (!headDimVStringAttr) {
+      return rewriter.notifyMatchFailure(
+          srcOp, "FlashMlaPrefill op requires head_dim_v attribute.");
+    }
+    uint32_t headDimV;
+    if (!llvm::to_integer(headDimVStringAttr.getValue(), headDimV) ||
+        headDimV == 0) {
```
Comment (mmanzoorTT) [inline lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp]: For positive number. 
```suggestion
    if (!llvm::to_integer(headDimVStringAttr.getValue(), headDimV) ||
        headDimV <= 0) {
```

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp [hunk 3]
```diff
@@ -8614,6 +8614,131 @@ class StableHLOToTTIRScaledDotProductAttentionOpConversionPattern
...
+            srcOp,
+            "scale attribute string must be convertible to float. Received \"" +
+                scaleStringAttr.getValue() + "\".");
+      }
+      scale = _scale;
+    }
+    FloatAttr scaleAttr =
+        scale ? rewriter.getF32FloatAttr(scale.value()) : nullptr;
+
+    // has_value / has_attention_mask flags.
```
Comment (mmanzoorTT) [inline lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp]: Suggestion. Move `has_attention_mask` comment down with `attention_mask` retrieval.

**Review by mmanzoorTT — COMMENTED** [review]
Stablehlo changes look good

**Review by sshonTT — APPROVED** [review]
looks good for sharding rule part

**Review by jgrimTT — APPROVED** [review]
I recommend adding a snippet to `tt-mlir/test/python/golden/mlir_snippets/ttir` to test the parse/split functionality, but not a blocker, looks good!

File: lib/Dialect/TTNN/Interfaces/TTNNOpModelInterface.cpp (hunk 1 above)
Comment (hshahTT) [reply lib/Dialect/TTNN/Interfaces/TTNNOpModelInterface.cpp]: Done

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp (hunk 2 above)
Comment (hshahTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp]: Done

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp (hunk 3 above)
Comment (hshahTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp]: Done

File: include/ttmlir/OpModel/TTNN/TTNNOpModel.h [hunk 4]
```diff
@@ -862,6 +862,33 @@ struct OpModel<ScaledDotProductAttentionOp> {
                TTNNLayoutAttr outputLayout);
 };
 
+//===----------------------------------------------------------------------===//
+// FlashMlaPrefillOp
+//===----------------------------------------------------------------------===//
+template <>
+struct OpModel<FlashMlaPrefillOp> {
+  static llvm::Expected<OpConstraints> getOpConstraints(
+      ttcore::GridAttr deviceGrid, llvm::ArrayRef<int64_t> queryShape,
```
Comment (rpavlovicTT) [inline include/ttmlir/OpModel/TTNN/TTNNOpModel.h]: deviceGrid not needed anymore.

**Review by rpavlovicTT — APPROVED** [review]

File: include/ttmlir/OpModel/TTNN/TTNNOpModel.h (hunk 4 above)
Comment (hshahTT) [reply include/ttmlir/OpModel/TTNN/TTNNOpModel.h]: Updated

**Review by mmanzoorTT — APPROVED** [review]

File: include/ttmlir/Dialect/TTIR/IR/TTIROps.td [hunk 5]
```diff
@@ -6187,6 +6187,48 @@ def TTIR_ScaledDotProductAttentionOp : TTIR_NamedOp<"scaled_dot_product_attentio
   let hasVerifier = 1;
 }
 
+def TTIR_FlashMlaPrefillOp : TTIR_NamedOp<"flash_mla_prefill", [AttrSizedOperandSegments]> {
```
Comment (sdjordjevicTT) [inline include/ttmlir/Dialect/TTIR/IR/TTIROps.td]: Can we, instead of a new TTIR op, use TTCore_CompositeOp?

**Review by sdjordjevicTT — COMMENTED** [review]
Looks good overall, one suggestion below.

File: include/ttmlir/Dialect/TTIR/IR/TTIROps.td (hunk 5 above)
Comment (sdjordjevicTT) [reply include/ttmlir/Dialect/TTIR/IR/TTIROps.td]: Example of composite usage:
https://github.com/tenstorrent/tt-mlir/pull/8717

Comment (hshahTT) [reply include/ttmlir/Dialect/TTIR/IR/TTIROps.td]: I'm unfamiliar with TTCore_CompositeOp. What's the benefit to using this instead of a TTIR op?

Comment (sdjordjevicTT) [reply include/ttmlir/Dialect/TTIR/IR/TTIROps.td]: We shouldn't model a TTIR dialect based on whatever is found in TTNN. TTIR should have a well-known set of ops, and some are very TTNN specific, like this one. For cases such as this, we introduced the `TTCore_CompositeOp` that is very similar to `stablehlo.composite`. So instead of lowering to the dedicated TTIR op, we should lower this op to TTCore_CompositeOp.

Comment (hshahTT) [reply include/ttmlir/Dialect/TTIR/IR/TTIROps.td]: I added that lowering in this commit: https://github.com/tenstorrent/tt-mlir/pull/8636/changes/0176a01c619e71159f5411e4490ff4fd7a6d28d5

Could you take a look again?

**Review by sdjordjevicTT — APPROVED** [review]
Looks great, it would be great if we could add some Python builder test that targets stablehlo custom call op.

File: lib/Conversion/TTNNToEmitPy/TTNNToEmitPy.cpp [hunk 6]
```diff
@@ -4375,6 +4375,50 @@ class ScaledDotProductAttentionOpConversionPattern
...
+                  ConversionPatternRewriter &rewriter) const override {
+
+    ttnn_to_emitpy::EmitPyTTNNEmitter<mlir::tt::ttnn::FlashMlaPrefillOp>
+        emitter(srcOp, adaptor, rewriter);
+
+    llvm::SmallVector<mlir::Attribute> args{
+        emitter.emit(srcOp.getQuery()),
+        emitter.emit(srcOp.getKey()),
+        emitter.emit(srcOp.getHeadDimV()),
+        emitter.emit(srcOp.getValue(), "input_tensor_v"),
```
Comment (amilovanovicTT) [inline lib/Conversion/TTNNToEmitPy/TTNNToEmitPy.cpp]: In the nanobind, this op is overloaded such that only one of `head_dim_v` and `input_tensor_v` is present, so emitting both of these one after another will crash. Could you extend the checks in the emitpy lit test to cover this properly?

File: tools/builder/stablehlo/stablehlo_builder.py [hunk 7]
```diff
@@ -235,6 +235,39 @@ def _get_location(self) -> Location:
...
+            key.float(),
+            value_t.float(),
+            attn_mask=attn_mask,
+            is_causal=effective_causal,
+            scale=scale,
+            enable_gqa=query.shape[1] != key.shape[1],
+        )
+
+        return output.to(output_dtype)
+
```
Comment (jgrimTT) [inline tools/builder/stablehlo/stablehlo_builder.py]: Is there a reason this golden was added to `stablehlo_builder.py` instead of `mapping.py`?

File: lib/Conversion/TTNNToEmitPy/TTNNToEmitPy.cpp (hunk 6 above)
Comment (hshahTT) [reply lib/Conversion/TTNNToEmitPy/TTNNToEmitPy.cpp]: fixed here: https://github.com/tenstorrent/tt-mlir/pull/8636/changes/63a1e14de447c290d1f4f4e9490f7ff1bfc0b6db . PTAL

File: tools/builder/stablehlo/stablehlo_builder.py (hunk 7 above)
Comment (hshahTT) [reply tools/builder/stablehlo/stablehlo_builder.py]: good point. I refactored it in this commit: https://github.com/tenstorrent/tt-mlir/pull/8636/changes/50766ef707d3c7d5464e03f4fb503c24ee67d0b5

**Review by jgrimTT — APPROVED** [review]
Thanks!

**Review by amilovanovicTT — APPROVED** [review]
